src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,48 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
+import mate.academy.exception.NoUserInCacheException;
+import mate.academy.exception.RequestProcessingException;
 
 public class AsyncRequestProcessor {
+    private static Map<String, UserData> cache = new ConcurrentHashMap<>();
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        return CompletableFuture.supplyAsync(
+                () -> findUserInCache(userId), executor).exceptionally(ex -> {
+                    try {
+                        UserData userData = emulateDataBaseAccess(userId);
+                        CompletableFuture.runAsync(() -> cacheFetchedData(userId, userData));
+                        return userData;
+                    } catch (InterruptedException e) {
+                        throw new RequestProcessingException("Failed to process "
+                                + "the request for userId=" + userId, e);
+                    }
+                });
+    }
+
+    private UserData findUserInCache(String userId) {
+        if (cache.containsKey(userId)) {
+            return cache.get(userId);
+        } else {
+            throw new NoUserInCacheException();
```

Comment:
The method `findUserInCache` throws a `NoUserInCacheException` which is not caught or declared to be thrown. This can cause the CompletableFuture to complete exceptionally without a clear recovery path. It's better to return an Optional or a special value to indicate the absence of the user in the cache.